### PR TITLE
Creates PP to VPP for Stochastic Tools module

### LINKED ISSUE #9419: VectorPostprocessor to Postprocessor transfer.
### Description of the enhancement or error report
Stochastic analysis needs means for transfering data. One method will be to have VectorPostprocessor on the master that can send/receive data to a sub apps postprocessor. The VPP will be able to connect to multiple PPs on the sub (i.e., a vector for each). The vectors will have a row for each sub-app.

### Rationale for the enhancement or information for reproducing the error
Required for stochastic runs and will be a good general purpose transfer.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
New feature, no impact.

### PR DESCRIPTION
**Pull request by aeslaughter**
(closes #9419)

### CODE REVIEW

File: framework/include/samplers/Sampler.h [hunk 1]
```diff
@@ -50,6 +50,24 @@ class Sampler : public MooseObject, public SetupInterface, public DistributionIn
...
+    {
+      this->sample = sample;
+      this->row = row;
+    }
+
+    /// Sample number (i.e., the index in the matrix)
+    unsigned int sample;
+
+    /// Row number for the given sample matrix
+    unsigned int row;
```
Comment (andrsd) [inline framework/include/samplers/Sampler.h]: member variables should start with `_`

File: framework/include/samplers/Sampler.h [hunk 2]
```diff
@@ -68,11 +86,33 @@ class Sampler : public MooseObject, public SetupInterface, public DistributionIn
...
+   * Return the sample names, by default 'sample_0, sample_1, etc.' is used.
+   *
+   * If custom names are required then utilize the setSampleNames method.
+   */
+  const std::vector<std::string> & getSampleNames() const { return _sample_names; }
+
+  /**
+   * Set the sample names.
+   */
+  void setSampleNames(std::vector<std::string> names);
```
Comment (andrsd) [inline framework/include/samplers/Sampler.h]: `const std::vector<std::string> &`

File: framework/include/samplers/Sampler.h [hunk 3]
```diff
@@ -68,11 +86,33 @@ class Sampler : public MooseObject, public SetupInterface, public DistributionIn
...
 
   /**
+   * Return the sample names, by default 'sample_0, sample_1, etc.' is used.
+   *
+   * If custom names are required then utilize the setSampleNames method.
+   */
+  const std::vector<std::string> & getSampleNames() const { return _sample_names; }
+
+  /**
+   * Set the sample names.
```
Comment (andrsd) [inline framework/include/samplers/Sampler.h]: add `@param names ...`

File: framework/include/samplers/Sampler.h [hunk 4]
```diff
@@ -68,11 +86,33 @@ class Sampler : public MooseObject, public SetupInterface, public DistributionIn
...
+  void setSampleNames(std::vector<std::string> names);
+
+  /**
    * Store the state of the MooseRandom generator so that new calls to getSamples will create
    * new numbers.
    */
   void execute();
 
+  /**
+   * Return the Sample::Location for the given multi app index.
```
Comment (andrsd) [inline framework/include/samplers/Sampler.h]: Add `@param global_index...`

File: framework/include/samplers/Sampler.h [hunk 5]
```diff
@@ -95,12 +135,20 @@ class Sampler : public MooseObject, public SetupInterface, public DistributionIn
    */
   void setNumberOfRequiedRandomSeeds(const std::size_t & number);
 
+  /**
+   * Reinitialize the offsets and row counts.
```
Comment (andrsd) [inline framework/include/samplers/Sampler.h]: Add `@param data...`

File: framework/include/samplers/Sampler.h [hunk 6]
```diff
@@ -111,6 +159,12 @@ class Sampler : public MooseObject, public SetupInterface, public DistributionIn
 
   /// Initial random number seed
   const unsigned int & _seed;
+
+  /// Data offsets for computing location based on global index
+  std::vector<unsigned int> _offsets;
+
+  /// Total number of rows
+  unsigned int _total_rows = 0;
```
Comment (andrsd) [inline framework/include/samplers/Sampler.h]: I am fine with doing this, it is just that I have not seen others doing it...

File: framework/src/samplers/Sampler.C [hunk 7]
```diff
@@ -86,3 +112,30 @@ Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
...
+}
+
+Sampler::Location
+Sampler::getLocation(unsigned int global_index)
+{
+  if (_offsets.empty())
+    reinit(getSamples());
+
+  std::vector<unsigned int>::const_iterator iter;
+  iter = std::upper_bound(_offsets.begin(), _offsets.end(), global_index) - 1;
```
Comment (andrsd) [inline framework/src/samplers/Sampler.C]: Does it look ugly after clang-format is done with, when you do it by assignment? No a big issue...

File: modules/stochastic_tools/include/vectorpostprocessors/StochasticResults.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+class StochasticResults : public GeneralVectorPostprocessor, SamplerInterface
+{
+public:
+  StochasticResults(const InputParameters & parameters);
+  void virtual initialize() override;
+  void virtual finalize() override {}
+  void virtual execute() override {}
+
+protected:
+  /// Initialize storage based on this
```
Comment (andrsd) [inline modules/stochastic_tools/include/vectorpostprocessors/StochasticResults.h]: `/** */` rather than `///`

File: modules/stochastic_tools/include/vectorpostprocessors/StochasticResults.h [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  void init(Sampler & _sampler);
+
+  /// Return the VectorPostprocessorValue for a given Sampler group index
```
Comment (andrsd) [inline modules/stochastic_tools/include/vectorpostprocessors/StochasticResults.h]: `/** */` rather than `///`

File: modules/stochastic_tools/src/multiapps/SamplerMultiApp.C [hunk 10]
```diff
@@ -5,9 +5,8 @@
 /*             See LICENSE for full restrictions                */
 /****************************************************************/
 
-// MOOSE includes
+// StochasticTools includes
```
Comment (andrsd) [inline modules/stochastic_tools/src/multiapps/SamplerMultiApp.C]: Personally, I would not bother commenting on #includes...

File: modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  if (!_results)
+    mooseError("The 'results' object must be a 'StochasticResults' object.");
+
+  _results->init(_sampler);
+}
+
+void
+SamplerPostprocessorTransfer::execute()
+{
+  // Number of values
```
Comment (andrsd) [inline modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C]: I do not understand why number of global apps == number of values, but that's probably because I am not familiar with your design.

**Review by andrsd — DISMISSED** [review]

File: framework/include/samplers/Sampler.h (hunk 1 above)
Comment (aeslaughter) [inline framework/include/samplers/Sampler.h]: Even for a struct that is just a custom location? 

I originally was using a std::pair, but using foo.first and foo.second wasn't as clear as foo.sample and foo.row.

I would like to avoid having to create methods for retrieving these indicies and also don't want to use foo._sample and foo._row.

File: framework/src/samplers/Sampler.C (hunk 7 above)
Comment (aeslaughter) [inline framework/src/samplers/Sampler.C]: I don't understand this question?

File: modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C (hunk 11 above)
Comment (aeslaughter) [inline modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C]: Each sub-app has a PP, which is being trasffered the master. So there is a row in the VPP for each sub-app.

File: framework/include/samplers/Sampler.h (hunk 1 above)
Comment (andrsd) [inline framework/include/samplers/Sampler.h]: > Even for a struct that is just a custom location?

We have `struct`s in MOOSE and their member variables start with `_`.

> I would like to avoid having to create methods for retrieving these indicies and also don't want to use foo._sample and foo._row.

If you do not want foo._row, foo._sample, then use methods. Seriously, how much code is this: ``` unsigned int sample() { return _sample; }```.

File: framework/src/samplers/Sampler.C (hunk 7 above)
Comment (andrsd) [inline framework/src/samplers/Sampler.C]: Normally, one would do:
```
std::vector<unsigned int>::const_iterator iter = std::upper_bound(_offsets.begin(), _offsets.end(), global_index) - 1;
```
i.e. on a single line. But, I am assuming clang-format will not keep it on a single line. So, was that a reason to have this on 2 lines?

File: modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C (hunk 11 above)
Comment (andrsd) [inline modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C]: Ok. If I missed this in your PR, then this is fine. Otherwise, you might want to include this somewhere, so people know why number of global apps == number of values.

File: framework/include/samplers/Sampler.h (hunk 6 above)
Comment (permcody) [inline framework/include/samplers/Sampler.h]: Yeah, I ran into problems with this the first time I tried this when we first switched. I can't remember what it was now. It's inconsistent with everything else but I suppose it's OK.

File: framework/src/samplers/Sampler.C [hunk 12]
```diff
@@ -86,3 +112,33 @@ Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
...
+}
+
+Sampler::Location
+Sampler::getLocation(unsigned int global_index)
+{
+  if (_offsets.empty())
+    reinit(getSamples());
+
+  std::vector<unsigned int>::const_iterator iter =
+      std::upper_bound(_offsets.begin(), _offsets.end(), global_index) - 1;
```
Comment (permcody) [inline framework/src/samplers/Sampler.C]: `- 1` is dangerous, this will fail if the upper_bound returns an iterator pointing to the first item. Also, if no item is found you'll get the pointer to the end which means it's not found, however you'll point back to the real last item in that case. If all this is fine, it should probably be asserted and/or documented.

File: modules/stochastic_tools/include/vectorpostprocessors/StochasticResults.h [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+  /// Storage for declared vectors
+  std::vector<VectorPostprocessorValue *> _sample_vectors;
+
+  /// The sampler to extract data
+  Sampler * _sampler = nullptr;
+
+  /// This object is designed to work with the SamplerPostprocessorTransfer, the init and
+  /// get methods should only be called by this class.
+  friend class SamplerPostprocessorTransfer;
```
Comment (permcody) [inline modules/stochastic_tools/include/vectorpostprocessors/StochasticResults.h]: Yes, but `friend` gives access to all the members, not just the methods you mention here. I'd prefer not to see `friend` if you can avoid it, is there really that much risk to making those members public? `execute` already is, `init` isn't but does it matter?

File: modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+  if (!_sampler_multi_app)
+    mooseError("The 'multi_app' must be a 'SamplerMultiApp.'");
+}
+
+void
+SamplerPostprocessorTransfer::initialSetup()
+{
+  const ExecuteMooseObjectWarehouse<UserObject> & user_objects = _fe_problem.getUserObjects();
+  std::shared_ptr<UserObject> uo = user_objects.getActiveObject(_results_name);
+  _results = std::dynamic_pointer_cast<StochasticResults>(uo).get();
```
Comment (permcody) [inline modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C]: Minor: This line of code is weird but OK. The `dynamic_pointer_cast` might fail but then you are calling `get` on it anyway. I checked the standard and this is OK, but it's a little odd. Since you are just throwing a MOOSE error if it fails, you could probably restructure this just a bit to clean it up.

File: modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+{
+  // Number of PP is equal to the number of MultiApps
+  const unsigned int n = _multi_app->numGlobalApps();
+
+  // Collect the PP values for this processor
+  std::vector<PostprocessorValue> values;
+  values.reserve(_multi_app->numLocalApps());
+  for (unsigned int i = 0; i < n; i++)
+  {
+    if (_multi_app->hasLocalApp(i))
```
Comment (permcody) [inline modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C]: If this fails, then the indexing between `values` and the multi_app numbering will be off. Is that OK? Maybe you should size the values vector (as opposed to reserving) and just index into it?

File: modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+  // Gather the PP values from all ranks
+  _communicator.allgather<PostprocessorValue>(values);
+
+  // Update VPP
+  for (unsigned int i = 0; i < n; i++)
+  {
+    Sampler::Location loc = _sampler.getLocation(i);
+    VectorPostprocessorValue & vpp = _results->getVectorPostprocessorValueByGroup(loc.sample());
+    vpp[loc.row()] = values[i];
```
Comment (permcody) [inline modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C]: See my note above, it's possible that the *i*th index is empty if any of the `hasLocalApp` calls fail.

File: framework/include/samplers/Sampler.h (hunk 6 above)
Comment (aeslaughter) [inline framework/include/samplers/Sampler.h]: Are you referring to MooseRandom with this statement? These parameters are unrelated to random number stuff, they are for indexing into the sampler data

File: modules/stochastic_tools/include/vectorpostprocessors/StochasticResults.h (hunk 13 above)
Comment (aeslaughter) [inline modules/stochastic_tools/include/vectorpostprocessors/StochasticResults.h]: I can change this, I don't care all that much.

File: modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C (hunk 14 above)
Comment (aeslaughter) [inline modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C]: This is a mistake, the .get should go on the line above and then just use a regular dynamic_cast

File: modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C (hunk 15 above)
Comment (aeslaughter) [inline modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C]: This is OK, it certainly does fail. This is a consequence of how the MultiApps are stored and accessed. The Transfer API is based on global index, thus you must loop all the global index and only do work if you are local.

> If this fails, then the indexing between values and the multi_app numbering will be off.

They are not supposed to be in agreement, each processors is populating the values, then the values are assembled into one vector.

```
[ [N0_0, N0_1, ..., N0_n], [N1_0, N1,_1, .. N1_n], ...]
```

This does assume that MultiApps are distributed in a continuous fashion, which I believe is the case.

File: framework/include/samplers/Sampler.h (hunk 6 above)
Comment (permcody) [inline framework/include/samplers/Sampler.h]: No, initializing a member variable in the .h file. It's a C++11 thing.

File: framework/src/samplers/Sampler.C [hunk 17]
```diff
@@ -86,3 +116,38 @@ Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
...
+  if (_offsets.empty())
+    reinit(getSamples());
+
+  mooseAssert(_offsets.size() > 1,
+              "The getSamples method returned an empty vector, if you are seeing this you have "
+              "done something to bypass another assert in the 'getSamples' method that should "
+              "prevent this message.");
+
+  std::vector<unsigned int>::const_iterator iter =
+      std::upper_bound(_offsets.begin(), _offsets.end(), global_index) - 1;
```
Comment (permcody) [inline framework/src/samplers/Sampler.C]: Why not std::lower_bound here? Then you wouldn't need a '- 1`.

File: framework/src/samplers/Sampler.C [hunk 18]
```diff
@@ -86,3 +116,38 @@ Sampler::setNumberOfRequiedRandomSeeds(const std::size_t & number)
...
+  return Sampler::Location(iter - _offsets.begin(), global_index - *iter);
```
Comment (permcody) [inline framework/src/samplers/Sampler.C]: `std::distance` is the more general way to perform these indexing calculations.

File: modules/stochastic_tools/include/transfers/SamplerPostprocessorTransfer.h [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+
+protected:
+  /// Name of VPP that will store the data
+  const VectorPostprocessorName & _results_name;
+
+  /// SamplerMultiApp that this transfer is working with
+  SamplerMultiApp * _sampler_multi_app;
+
+  /// Sampler object that is retrieved from the SamplerMultiApp
+  Sampler & _sampler;
```
Comment (permcody) [inline modules/stochastic_tools/include/transfers/SamplerPostprocessorTransfer.h]: Looks like this PR always works in "from_multiapp" mode. In that case should the source object be `const`?

**Review by permcody — COMMENTED** [review]
Just one or two more nitpicks

File: modules/stochastic_tools/include/transfers/SamplerPostprocessorTransfer.h (hunk 19 above)
Comment (aeslaughter) [inline modules/stochastic_tools/include/transfers/SamplerPostprocessorTransfer.h]: no because calling getSamples is not a const operation

File: framework/src/samplers/Sampler.C (hunk 17 above)
Comment (aeslaughter) [inline framework/src/samplers/Sampler.C]: std::lower uses not less, which when equal does the wrong thing. std::upper uses greater than, so the equal case can't happen.

Comment (permcody) [inline framework/src/samplers/Sampler.C]: OK - that's complicated and should definitely be documented here.

**Comment by permcody** [comment]
@aeslaughter - This will have to be rebased before we can merge.

**Comment by aeslaughter** [comment]
@permcody This is ready to go.

**Comment by permcody** [comment]
I looked through your addressing comments commit and still see several unaddressed comments. Can you take another look before I review?

**Comment by aeslaughter** [comment]
@permcody I think I have them addressed now, I lost some changes in a rebase or something else stupid I did.

**Comment by permcody** [comment]
I still don't see anything different since you commented yesterday. Did you forget to push?

File: framework/src/samplers/Sampler.C (hunk 17 above)
Comment (permcody) [reply framework/src/samplers/Sampler.C]: @aeslaughter - I'm still having a hard time with this. We routinely use lower_bound, so much so that I just put in a macro yesterday to replace the pattern of using lower bound to get the iterator to the right spot for insertions to occur (i.e. if the item is missing, it'll point where the item is supposed to go, if the item is present, it'll point at the item). Not less is used because that's the standard operator you need to define to have items stored in any sorted container (set or map). Greater than, is not normally required. I'm not convinced that this is correct because I think it misses edge cases. Can you provide a counter example?

Comment (permcody) [reply framework/src/samplers/Sampler.C]: Lower bound usage in MOOSE: https://github.com/idaholab/moose/commit/6580d8254026826e6c0c08c78b2111612284c39b

Comment (aeslaughter) [reply framework/src/samplers/Sampler.C]: I added a unit test that shows the behavior I need. I do not need an insert location, I am looking for a starting index based on an offset. In the example below, given any global index in need the index that gives me the starting point of that subset.

```
GLOBAL:     0     1      2     3        4      5     6      7     8
VECTORS: [x_0, x_1, x_2, x_3], [y_0, y_1, y_2, y_3, y_4,...]
```

Comment (permcody) [reply framework/src/samplers/Sampler.C]: Thanks, I see now.

**Review by permcody — APPROVED** [review]
